lib/memoist.rb

```diff
@@ -63,28 +63,49 @@ def unmemoize_all
       flush_cache
     end
 
+    def init_lazy_memoized_structs(names)
+      structs = self.class.memoized_methods
+      return structs if names.empty?
+
+      structs.select { |s| names.include?(s.memoized_method) }
+    end
+
     def prime_cache(*method_names)
-      method_names = self.class.memoized_methods if method_names.empty?
-      method_names.each do |method_name|
-        if method(Memoist.unmemoized_method_for(method_name)).arity == 0
-          __send__(method_name)
+      memoized_structs = init_lazy_memoized_structs(method_names)
+      memoized_structs.each do |memoized_struct|
+        if method(memoized_struct.unmemoized_method).arity == 0
+          __send__(memoized_struct.memoized_method)
         else
-          ivar = Memoist.memoized_ivar_for(method_name)
-          instance_variable_set(ivar, {})
+          instance_variable_set(memoized_struct.ivar, {})
         end
       end
     end
 
     def flush_cache(*method_names)
-      method_names = self.class.memoized_methods if method_names.empty?
-
-      method_names.each do |method_name|
-        ivar = Memoist.memoized_ivar_for(method_name)
-        remove_instance_variable(ivar) if instance_variable_defined?(ivar)
+      memoized_structs = init_lazy_memoized_structs(method_names)
+      memoized_structs.each do |memoized_struct|
+        remove_instance_variable(memoized_struct.ivar) if instance_variable_defined?(memoized_struct.ivar)
       end
     end
   end
 
+  MemoizedMethodStruct = Struct.new(:memoized_method, :ivar, :unmemoized_method) do
+    def initialize(*args)
+      super
+      @memoized_method = args[0]
+      @ivar = args[1]
+      @unmemoized_method = args[2]
+    end
+
+    def unmemoized_method
+      @unmemoized_method ||= Memoist.unmemoized_method_for(@memoized_method, nil)
```

Comment:
The previous implementation of flush/prime cache would call these methods a lot.  They didn't seem to belong in those methods so a call to the `unmemoized_method_for` or `memoized_ivar_for` that is only evaluated once per memoized method on first call to the method seemed like the better way.  Also, tucking them away in this Struct/class seemed to be a better place than flush/prime cache.